proofs/signatures/er_test.plf

```diff
@@ -0,0 +1,116 @@
+; Depends on er.plf
+
+; Type for representing success when testing a side-condition
+(declare TestSuccess type)
+
+; Test for clause_add_suffix_all
+(declare test_clause_add_suffix_all
+         (! list clause
+            (! suffix clause
+               (! result cnf
+                  (! (^ (clause_add_suffix_all list suffix) result)
+                     TestSuccess)))))
+
+(check
+  (% a lit
+  (% b lit
+  (% c lit
+     (test_clause_add_suffix_all
+       (clc a (clc b cln))
+       (clc c cln)
+       (cnfc (clc a (clc c cln))
+        (cnfc (clc b (clc c cln))
+        cnfn))
+     )
+  )))
+)
+
+; This is a circuitous proof that uses the definition introduction and
+; unrolling rules
+(check
+  (% v1 var
+  (% v2 var
+    (% pf_c1 (holds (clc (pos v1) (clc (pos v2) cln)))
+    (% pf_c2 (holds (clc (neg v1) (clc (pos v2) cln)))
+    (% pf_c3 (holds (clc (pos v1) (clc (neg v2) cln)))
+    (% pf_c4 (holds (clc (neg v1) (clc (neg v2) cln)))
+       (: (holds cln)
+          (decl_definition
+            (neg v1)
+            (clc (neg v2) cln)
+            (\ v3
```

Comment:
Consider the following variant of this example:
```
; This is a circuitous proof that uses the definition introduction and
; unrolling rules
(check
  (% v1 var
  (% v2 var
  (% v3 var
    (% pf_dummy (holds (clc (pos v3) (clc (neg v3) cln)))
    (% pf_c1 (holds (clc (pos v1) (clc (pos v2) cln)))
    (% pf_c2 (holds (clc (neg v1) (clc (pos v2) cln)))
    (% pf_c3 (holds (clc (pos v1) (clc (neg v2) cln)))
    (% pf_c4 (holds (clc (neg v1) (clc (neg v2) cln)))
       (: (holds cln)
          (decl_definition
            (neg v1)
            (clc (neg v2) cln)
            (\ v3
            (\ def3
              (clausify_definition _ _ _ def3 _ _
                (\ pf_c5 ; type: (holds (clc (pos v3) (clc (neg v2) cln)))
                (\ pf_c6 ; type: (holds (clc (pos v3) (clc (pos v1) cln)))
                (\ pf_cnf3 ; type: (cnf_holds (cnfc (clc (neg v2) (clc (neg v3) (clc (neg v1) cln))) cnfn))
                  (cnfc_unroll_towards_bottom _ _ pf_cnf3
                    (\ pf_c7 ; type: (clc (neg v2) (clc (neg v3) (clc (neg v1) cln)))
                    (\ pf_cnfn
                                                                          ; Clauses
                  (satlem_simplify _ _ _ (R _ _ pf_c1 pf_c2 v1)  (\ pf_c8 ; v2
                  (satlem_simplify _ _ _ (R _ _ pf_c8 pf_c5 v2)  (\ pf_c9 ; v3
                  (satlem_simplify _ _ _ (R _ _ pf_c9 pf_c7 v3)  (\ pf_c10 ; ~v2  ~v1
                  (satlem_simplify _ _ _ (Q _ _ pf_c10 pf_c8 v2)  (\ pf_c11 ; ~v1
                  (satlem_simplify _ _ _ (Q _ _ pf_c11 pf_c3 v1)  (\ pf_c12 ; ~v2
                  (satlem_simplify _ _ _ (Q _ _ pf_c12 pf_c8 v2)  (\ pf_c13 ; empty
                        pf_c13
                  ))
                  ))
                  ))
                  ))
                  ))
                  ))
                  )))
                )))
              )
            ))
          )
        )
    ))))
  ))))
)
```

The only difference is that I added a dummy `v3` at the beginning. This proof still type-checks, which is good, but I am wondering, can't a similar trick can be used in order to make a bad proof type-check?
Can we (and should we) somehow verify that the fresh variable in the sense of the proof is also fresh in the sense of the lfsc `check` clause?

Reply:
This is a good question. I assume that LFSC has lexical scoping with shadowing, but I don't **know** that.

If it doesn't have proper lexical scoping with shadowing then we should probably fix that. On the other hand, maybe there is some other language mechanism that could be used to implement freshness! I can't think of one right now though.